lib/byron/src/Cardano/Wallet/Byron/Network.hs

```diff
@@ -0,0 +1,611 @@
+    , ClientStNext (..)
+    , chainSyncClientPeer
+    )
+import Ouroboros.Network.Protocol.ChainSync.Codec
+    ( codecChainSync )
+import Ouroboros.Network.Protocol.ChainSync.Type
+    ( ChainSync )
+import Ouroboros.Network.Protocol.Handshake.Version
+    ( DictVersion (..), simpleSingletonVersions )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Client
+    ( LocalTxSubmissionClient (..), localTxSubmissionClientPeer )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Codec
+    ( codecLocalTxSubmission )
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type
+    ( LocalTxSubmission )
+
+import qualified Cardano.Wallet.Primitive.Types as W
+import qualified Codec.Serialise as CBOR
+import qualified Data.Text as T
+import qualified Network.Socket as Socket
+
+withNetworkLayer
+    :: forall a. ()
+    => Trace IO Text
+        -- ^ Logging of network layer startup
+    -> W.BlockchainParameters
+        -- ^ Static blockchain parameters
+    -> AddrInfo
+        -- ^ Socket for communicating with the node
+    -> (NodeToClientVersionData, CodecCBORTerm Text NodeToClientVersionData)
+        -- ^ Codecs for the node's client
+    -> (NetworkLayer IO Byron ByronBlock -> IO a)
+        -- ^ Action to run with the network layer.
+    -> IO a
+withNetworkLayer tr bp addrInfo versionData action = do
+    queue <- atomically newTQueue
+
+    let client = OuroborosInitiatorApplication $ \pid -> \case
+            ChainSyncWithBlocksPtcl ->
+                let tr' = contramap (T.pack . show) $ trMessage tr in
+                chainSyncWithBlocks tr' pid (W.getGenesisBlockHash bp) queue
+            LocalTxSubmissionPtcl ->
+                localTxSubmission nullTracer pid
+
+    bracket (before client) after (inBetween queue)
+  where
+    before client =
+        forkIO $ connectClient client versionData addrInfo
+    inBetween queue _ =
+        action (newNetworkLayer bp queue)
+    after =
+        killThread
+
+-- | Create an instance of the network layer, using the given communication
+-- channel.
+newNetworkLayer
+    :: MonadSTM m
+    => W.BlockchainParameters
+        -- ^ Static blockchain parameters
+    -> TQueue m (NetworkClientCmd m)
+        -- ^ Communication channel with the Network Client
+    -> NetworkLayer m Byron ByronBlock
+newNetworkLayer bp queue = NetworkLayer
+    { networkTip = _networkTip
+    , findIntersection = _findIntersection
+    , nextBlocks = _nextBlocks
+    , initCursor = _initCursor
+    , cursorSlotId = _cursorSlotId
+    , postTx = _postTx
+    , staticBlockchainParameters = _staticBlockchainParameters
+    , stakeDistribution = _stakeDistribution
+    , getAccountBalance = _getAccountBalance
+    }
+  where
+    _initCursor headers = do
+        let points = genesisPoint : (toPoint <$> headers)
+        queue `send` CmdFindIntersection points >>= \case
+            Left e  -> throw e
+            Right intersection -> pure $ Cursor intersection
+
+    _nextBlocks _ = do
```

Comment:
The `_` would mean:
- `nextBlocks t >> nextBlocks t` doesn't work as you'd expect
- `RetryImmediately :: FollowAction` will silently not work as expected when passed to `follow`

🤷‍♂

Reply:
I am not sure to follow you on this ?

Reply:
Ah, yes. The protocol is stateful now indeed. So using the same cursor twice will have an unexpected behavior. I'll see if we can handle that properly.

Reply:
Added a note.

Reply:
I feels much better to me now when you put the queue inside the `Cursor` 👍 

One can look at the cursor, see that there is some kind of state there, and not be surprised.

Also, `initCursor` will always create a new node-connection & cursor. This is neat. Had the state been in the NetworkLayer, we could have accidentally messed up an ongoing chain-sync by calling `initCursor` in the middle of it (which would have changed the node's notion of where the wallet's tip). I imagine at least.